core/src/main/scala/shapeless/ops/hlists.scala

```diff
@@ -2036,20 +2044,24 @@ object hlist {
      */
     override type Aux[P <: HList, S <: HList, Out0 <: HList] = Prepend[P, S] { type Out = Out0 }
 
-    implicit def hnilPrepend0[P <: HList, S <: HNil]: Aux[P, S, P] =
+    implicit def hnilPrepend0[P <: HList, S <: HNil](implicit n: HNil.type <:< S): Aux[P, S, P] =
       new Prepend[P, S] {
         type Out = P
         def apply(prefix : P, suffix : S): P = prefix
+
+        def inverse(p: P): (P, S) = (p, n(HNil))
       }
   }
 
   object Prepend extends LowPriorityPrepend {
     def apply[P <: HList, S <: HList](implicit prepend: Prepend[P, S]): Aux[P, S, prepend.Out] = prepend
 
-    implicit def hnilPrepend1[P <: HNil, S <: HList]: Aux[P, S, S] =
+    implicit def hnilPrepend1[P <: HNil, S <: HList](implicit n: HNil.type <:< P): Aux[P, S, S] =
```

Comment:
Same comment about the type parameter.
Let's at least try it out and see if all tests pass.

Reply:
Similar problem as above:
```
shapeless/core/src/test/scala/shapeless/hlist.scala:426:55: could not find implicit value for parameter prepend: shapeless.ops.hlist.Prepend[shapeless.HNil.type,L]
```